discovery/sync_manager.go

```diff
@@ -0,0 +1,640 @@
+		pendingActiveSyncerQueue: list.New(),
+		pendingActiveSyncers:     make(map[routing.Vertex]*activeSyncer),
+		newActiveSyncers:         make(chan *activeSyncer),
+		staleActiveSyncers:       make(chan routing.Vertex),
+		quit:                     make(chan struct{}),
+	}
+}
+
+// Start starts the SyncManager in order to properly carry out its duties.
+func (m *SyncManager) Start() {
+	m.start.Do(func() {
+		m.wg.Add(1)
+		go m.syncerHandler()
+	})
+}
+
+// Stop stops the SyncManager from performing its duties.
+func (m *SyncManager) Stop() {
+	m.stop.Do(func() {
+		close(m.quit)
+		m.wg.Wait()
+
+		for _, syncer := range m.activeSyncers {
+			syncer.Stop()
+		}
+		for _, syncer := range m.inactiveSyncers {
+			syncer.Stop()
+		}
+	})
+}
+
+// enqueueActiveSyncer adds the given gossip syncer to the end of the
+// round-robin queue.
+func (m *SyncManager) enqueueActiveSyncer(s *GossipSyncer, transition bool) {
+	select {
+	case m.newActiveSyncers <- &activeSyncer{
+		GossipSyncer: s,
+		transition:   transition,
+	}:
+	case <-m.quit:
+	}
+}
+
+// dequeueActiveSyncer removes the syncer for the given peer from the
+// round-robin queue.
+func (m *SyncManager) dequeueActiveSyncer(peer routing.Vertex) {
+	select {
+	case m.staleActiveSyncers <- peer:
+	case <-m.quit:
+	}
+}
+
+// syncerHandler is the SyncManager's main event loop responsible for:
+//
+//   1. Finding new peers to receive graph updates from to ensure we don't only
+//   receive them from the same set of peers.
+//
+//   2. Finding new peers to force a full sync with to ensure we have as much of
+//   the public network as possible.
+//
+//   3. Managing the round-robin queue of our active syncers to ensure they
+//   don't overlap and request the same set of channels, which significantly
+//   reduces bandwidth usage.
+//
+// NOTE: This must be run as a goroutine.
+func (m *SyncManager) syncerHandler() {
+	defer m.wg.Done()
+
+	// We'll start all of our relevant tickers, except for our
+	// ActiveSyncerTimeoutTicker, which should only be started when we start
+	// an activeSyncer.
+	m.cfg.RotateTicker.Resume()
+	defer m.cfg.RotateTicker.Stop()
+
+	m.cfg.HistoricalSyncTicker.Resume()
+	defer m.cfg.HistoricalSyncTicker.Stop()
+
+	defer m.cfg.ActiveSyncerTimeoutTicker.Stop()
+
+	var (
+		// current will hold the current activeSyncer we're waiting for
+		// to complete its state machine.
+		current *activeSyncer
+
+		// startNext will be responsible for containing the signal of
+		// when the current activeSyncer has completed its state
+		// machine. This signal allows us to start the next pending
+		// activeSyncer, if any.
+		startNext chan struct{}
+	)
+
+	// startSyncer is a helper closure we'll use to start a given syncer. If
+	// necessary, a sync transition to an ActiveSync sync type will be
+	// performed to ensure we receive new graph updates from this peer.
+	startSyncer := func(s *activeSyncer) error {
+		if s.transition {
```

Comment:
Is this `transition` bool really needed? We could rather start all syncers in the passive state, and transition them always to active when using them.